File: client-core/src/main/java/dole/card/SmartCard.java
package dole.card;

public interface SmartCard {
    void connect() throws Exception;
    void disconnect();

    default boolean isConnected() { return false; }
    boolean isPinSet() throws Exception;
    boolean isGenesisDone() throws Exception;
    boolean verifyPin(byte[] pin) throws Exception;
    boolean changePin(byte[] newPin) throws Exception;
    byte[] getPublicKey() throws Exception;
    byte[] getCertificate() throws Exception;
    byte[] processGenesis() throws Exception;
    byte[] processMint(byte[] payload) throws Exception;
    byte[] processBurn(byte[] payload) throws Exception;
    byte[] processSend(byte[] payload) throws Exception;
    void processReceive(byte[] payload) throws Exception;
    void addPeer(byte[] payload) throws Exception;
}

File: composeApp/src/commonMain/kotlin/dole/card/SmartCard.kt
package dole.card

interface SmartCard {
    fun connect()
    fun disconnect()
    val isConnected: Boolean
    val isMinter: Boolean
    val isPinSet: Boolean
    val isGenesisDone: Boolean
    val pinRetries: Int
    val publicKey: ByteArray?
    val certificate: ByteArray?

    fun verifyPin(pin: ByteArray): Boolean
    fun changePin(newPin: ByteArray): Boolean
    fun processGenesis(): ByteArray
    fun processMint(payload: ByteArray): ByteArray
    fun processBurn(payload: ByteArray): ByteArray
    fun processSend(payload: ByteArray): ByteArray
    fun processReceive(payload: ByteArray)
    fun addPeer(payload: ByteArray)
}